### PR TITLE
To master

### PR DESCRIPTION
**Pull request by EliorDayari**
20,26-3.

### CODE REVIEW

File: ShenkinStore/Controllers/TransactionsController.cs [hunk 1]
```diff
@@ -33,10 +38,9 @@ public async Task<IActionResult> Details(int? id)
                 return NotFound();
             }
 
-            var transaction = await _context.Transaction
-                .Include(t => t.Product)
-                .Include(t => t.User)
+            var transaction = await db.Transactions.Include(p => p.productslist)
```
Comment (EliorDayari) [inline ShenkinStore/Controllers/TransactionsController.cs]: This is the "join"

File: ShenkinStore/Controllers/ShoppingCartController.cs [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+                };
+                return View(viewModel);
+            }
+            else
+            {
+                return RedirectToAction("Index", "Home");
+            }
+        }
+
+
```
Comment (EliorDayari) [inline ShenkinStore/Controllers/ShoppingCartController.cs:44]: The new cart with the hash map

File: ShenkinStore/Controllers/ShoppingCartController.cs [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+        {
+            var userID = HttpContext.Session.GetString("UserID");
+            if (userID != null)
+            {
+                var addedProduct = db.Products.Single(product => product.ProductId == id);
+                var cart = ShoppingCart.GetCart(userID.ToString());
+
+
+                cart.AddToCart(addedProduct, amount);
+
```
Comment (zeevig) [inline ShenkinStore/Controllers/ShoppingCartController.cs:54]: Remove all the non relevant spaces

File: ShenkinStore/Controllers/ShoppingCartController.cs [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+                return RedirectToAction("Index", "ShoppingCart");
+            }
+            else
+            {
+                return RedirectToAction("Login", "Users");
+            }
+        }
+
+
+        public ActionResult IncreaseProduct(int id)
```
Comment (zeevig) [inline ShenkinStore/Controllers/ShoppingCartController.cs]: Change the name to something more understandable like (IncreaseProductAmount or something like that)

File: ShenkinStore/Controllers/ShoppingCartController.cs [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+            var userID = HttpContext.Session.GetString("UserID");
+
+            var cart = ShoppingCart.GetCart(userID.ToString());
+            cart.IncreaseAmount(id);
+            return RedirectToAction("Index", "ShoppingCart");
+
+
+
+        }
+        public ActionResult DecreaseProduct(int id)
```
Comment (zeevig) [inline ShenkinStore/Controllers/ShoppingCartController.cs]: Here as well

File: ShenkinStore/Controllers/ShoppingCartController.cs [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+            {
+                return RedirectToAction("Login", "Users");
+            }
+        }
+
+
+
+
+
+        /* Maybe We will need it on Future
```
Comment (zeevig) [inline ShenkinStore/Controllers/ShoppingCartController.cs]: Remove those comments

File: ShenkinStore/Controllers/ShoppingCartController.cs [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+            else
+            {
+                return RedirectToAction("Login", "Users");
+            }
+        }
+
+
+
+
+        //Kind of Grabge Collector, unmanneged
```
Comment (zeevig) [inline ShenkinStore/Controllers/ShoppingCartController.cs:155]: When you are using it?

File: ShenkinStore/Controllers/ShoppingCartController.cs [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+            {
+
+                var cart = ShoppingCart.GetCart(userID.ToString());
+                var cartItems = cart.GetCartItems();
+                foreach (var cartItem in cartItems)
+                {
+
+                    cart.RemoveFromCart(cartItem);
+                }
+                db.SaveChanges();
```
Comment (zeevig) [inline ShenkinStore/Controllers/ShoppingCartController.cs:118]: Why we have any connection to DB here?

File: ShenkinStore/Controllers/ShoppingCartController.cs [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+using ShenkinStore.Models;
+using ShenkinStore.ViewModels;
+
+namespace ShenkinStore.Controllers
+{
+    public class ShoppingCartController : Controller
+    {
+        ShenkinContext db = new ShenkinContext();
```
Comment (zeevig) [inline ShenkinStore/Controllers/ShoppingCartController.cs:14]: Why we are using the DB here?

File: ShenkinStore/Controllers/ShoppingCartController.cs [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+
+        public ActionResult emptyCart()
+        {
+            var userID = HttpContext.Session.GetString("UserID");
+            if (userID != null)
+            {
+
+                var cart = ShoppingCart.GetCart(userID.ToString());
+                var cartItems = cart.GetCartItems();
+                foreach (var cartItem in cartItems)
```
Comment (zeevig) [inline ShenkinStore/Controllers/ShoppingCartController.cs:113]: I think Maps have RevmoveAll built in so add a function in cart that does remove all and replace with this foreach

File: ShenkinStore/Controllers/TransactionsController.cs [hunk 11]
```diff
@@ -46,28 +50,35 @@ public async Task<IActionResult> Details(int? id)
         }
 
         // GET: Transactions/Create
-        public IActionResult Create()
+        // Using filter to allow access only to login users.
+        //[Authorize] - TODO: uncomment before you go live
```
Comment (zeevig) [inline ShenkinStore/Controllers/TransactionsController.cs:61]: Delete those comments

File: ShenkinStore/Controllers/ShoppingCartController.cs (hunk 8 above)
Comment (zeevig) [reply ShenkinStore/Controllers/ShoppingCartController.cs:118]: And if we need it I think it's better to use await with async save

File: ShenkinStore/Controllers/TransactionsController.cs [hunk 12]
```diff
@@ -46,28 +50,35 @@ public async Task<IActionResult> Details(int? id)
...
         [ValidateAntiForgeryToken]
-        public async Task<IActionResult> Create([Bind("TransactionId,TransactionDate,UserID,ProductID")] Transaction transaction)
+        public async Task<IActionResult> Create([Bind("TransactionId,Delivery,Paid,TransDate,Amount")] Transaction transaction)
         {
             if (ModelState.IsValid)
             {
-                _context.Add(transaction);
-                await _context.SaveChangesAsync();
-                return RedirectToAction(nameof(Index));
+               // db.Add(transaction);
```
Comment (zeevig) [inline ShenkinStore/Controllers/TransactionsController.cs]: Delete those comments

File: ShenkinStore/Models/ShenkinContext.cs [hunk 13]
```diff
@@ -15,16 +15,21 @@ public ShenkinContext(DbContextOptions<ShenkinContext> options)
...
-        //                optionsBuilder.UseSqlServer("Server=(localdb)\\MSSQLLocalDB;Database=Shenkin;Trusted_Connection=True;");
-        //            }
-        //        }
+        public DbSet<Product> Products { get; set; }
+        public DbSet<User> Users { get; set; }
+
+        public DbSet<Transaction> Transactions { get; set; }
+
+
+        //************* "A Comment in case we'll need for future use" *************
```
Comment (zeevig) [inline ShenkinStore/Models/ShenkinContext.cs]: Delete this comment

File: ShenkinStore/Models/ShoppingCart.cs [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+    public class ShoppingCart
+    {
+        // Database initiatl
+        ShenkinContext db = new ShenkinContext();
+        // Data for Recognation
+        public string ShoppingCartId { get; set; }
+
+        public const String CartSessionKey = "cartId";
+
+       public static Dictionary<int, int> productsAmount = new Dictionary<int, int>();
```
Comment (zeevig) [inline ShenkinStore/Models/ShoppingCart.cs:20]: I all those can be in 1 dictionary.
Dictionary<userId(int), Dictionary<Product, amount(int)>>.
And if you want the total just get all the prices of the products in a specific cart with lamda (if you need help LMK).

**Review by zeevig — COMMENTED** [review]
Please take a look at the comments

**Review by EliorDayari — COMMENTED** [review]
fixed

**Review by zeevig — APPROVED** [review]

**Pull request merged**